### PR TITLE
Add GPU accelerated tree construction plugin

### PR DESCRIPTION
**Pull request by RAMitchell**
Newer version of the algorithm supporting trees of arbitrary depth while keeping speed advantages.

All .cuh files should pass the linter now also.

### CODE REVIEW

**Comment by tqchen** [comment]
Thanks! I can merge this in.

Before doing that, I would like to discuss how can some of the files in the .cuh file could be shared with the original source code, especially the data structures used for loss calculations.

Here is rough way to do it
- Add a macro like XGB_DEVICE which will redirect to __host__ __device__ when compiled with nvcc
   - Can be detected with __NVCC__ macro

**Comment by tqchen** [comment]
@RAMitchell please let me know about your thoughts. We can also do that as a second PR that as a goal for refactoring,

**Comment by tqchen** [comment]
Also it would be nice if you could comment briefly on algorithm and data structure

- Do you keep a DMatrix cache on GPU, does copy need to happen in each iteration?

What changes to XGBoost's data structure could help make GPU computation easier

Thanks for the great job

**Comment by RAMitchell** [comment]
With regard to the loss functions do you mean changing the loss functions within the original xgboost code with a macro to avoid duplicating these functions in .cuh files? Or some more comprehensive generalisation of the code/data structures used? 

If the first then I can update the PR today or tomorrow easily enough, if the second I would recommend we discuss more and do a second PR. I'm quite interested to get some people using the code to pick up bugs and get feedback.

The algorithm will be the subject of my masters dissertation so I hope to release a paper before the end of the year.

The DMatrix is cached on the gpu between iterations - only new gpairs need to be copied. You will notice the first iteration is considerably slower as this data structure is created.

The requirements of the data structure on the GPU are quite different so I have created something new  and copied the DMatrix into it. 

If there was a device Dmatrix type it might be possible to produce a more complete GPU solution e.g. with files being copied directly to device memory, sorted on the device, gpairs calculated on the device, as opposed to just tree construction. I would have to run some benchmarks on other parts of the xgboost algorithm to know if this is worth doing. It could make xgboost more extensible for GPUs.

Happy to discuss more over skype if that suits.

**Comment by tqchen** [comment]
How about you do a bit changes to the first point to enable a bit more reuse and I can merge this in. I would be more than happy to discuss on what can be improved in general in terms of GPU computing.

**Comment by RAMitchell** [comment]
A brief description of the algorithm used:

Decision tree algorithms for numeric data essentially perform a scan operation from left to right  to enumerate through all possible splits. This scan operation is sometimes known as parallel prefix sum in GPU programming. It is very efficient on GPUs. See here for an explanation: 
http://http.developer.nvidia.com/GPUGems3/gpugems3_ch39.html

I use two separate algorithms for different tree depths. For the first six levels I enumerate through all fvalues while they are interleaved with respect to the tree node that currently reside in. This requires  a variation of scan sometimes called "multiscan". 

This is no longer possible above around depth six as the number of nodes exceeds 2^6 and the GPU  runs out of  temporary storage. After this all items are sorted into node buckets at each level to be scanned.

**Comment by tqchen** [comment]
Good, one final thing. I noticed that the indentation in the .cuh file are still set to 4 space. Do you mind to change them to .h file and do a bit of reformatting to  make it consistent with rest of code style. 

I will merge it in after it is done. 

As a second note, you are more than welcomed(actually encouraged) to send an tutorial about GPU based tree to xgboost tutorial, as well as DMLC's official blogpost. So that more people can aware of it and try it out.

**Comment by tqchen** [comment]
also please fix the lint check https://travis-ci.org/dmlc/xgboost/jobs/169076687

**Comment by tqchen** [comment]
Thanks. This is merged
